File: java-retrofit/src/main/java/com/iproov/androidapiclient/javaretrofit/TokenRequestSerializer.java
package com.iproov.androidapiclient.javaretrofit;

import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.google.gson.JsonSerializationContext;
import com.google.gson.JsonSerializer;

import java.lang.reflect.Type;
import java.util.Map;
/*
    This is only used to inline the options hashmap, gson default behaviour is to nest
 */
public class TokenRequestSerializer implements JsonSerializer<TokenRequest> {

    @Override
    public JsonElement serialize(TokenRequest src, Type typeOfSrc, JsonSerializationContext context) {
        JsonObject jsonObject = new JsonObject();

        // TODO This isn't scalable but using context.serialize with src results in an infinite loop
        jsonObject.addProperty("api_key", src.getApiKey());
        jsonObject.addProperty("secret", src.getSecret());
        jsonObject.addProperty("resource", src.getResource());
        jsonObject.addProperty("user_id", src.getUserId());
        jsonObject.addProperty("client", src.getClient());
        jsonObject.addProperty("assurance_type", src.getAssuranceType());

        Map<String, Object> options = src.getOptions();

        if (options == null)
            return jsonObject;

        JsonObject jsonOptions = context.serialize(options).getAsJsonObject();

        for (Map.Entry<String, JsonElement> entry : jsonOptions.entrySet()) {
            jsonObject.add(entry.getKey(), entry.getValue());
        }

        return jsonObject;
    }
}

File: kotlin-retrofit/src/main/java/com/iproov/androidapiclient/kotlinretrofit/TokenRequestSerializer.kt
package com.iproov.androidapiclient.kotlinretrofit

import com.google.gson.JsonElement
import com.google.gson.JsonObject
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer
import java.lang.reflect.Type

class TokenRequestSerializer : JsonSerializer<GetTokenRequest> {

    override fun serialize(src: GetTokenRequest?, typeOfSrc: Type?, context: JsonSerializationContext?): JsonElement =
        JsonObject().apply {
            // TODO This isn't scalable but using context.serialize with src results in an infinite loop
            addProperty("api_key", src!!.apiKey)
            addProperty("secret", src.secret)
            addProperty("resource", src.resource)
            addProperty("assurance_type", src.assuranceType)
            addProperty("user_id", src.userId)
            addProperty("riskProfile", src.riskProfile)
        }
}